cluster_utils/slurm_cluster_system.py

```diff
@@ -378,44 +449,23 @@ def mark_failed_jobs(self, jobs: Sequence[Job]) -> None:
         output = proc.stdout.decode()
         logger.debug("Output of sacct:\n%s", output)
 
-        # Output looks like this:
-        #
-        #     4597753|cpu-short|martius|2|FAILED|1:0
-        #     4597753.batch||martius|2|FAILED|1:0
-        #     4597753.extern||martius|2|COMPLETED|0:0
-        #     ...
-        #
-        # The ExitCode field has the format {exit_code}:{signal_that_killed_job_if_any}
-
-        for line in output.splitlines():
-            job_id, node_list, state, exit_code = line.split("|")
-
-            # Only check the line where job_id matches exactly the expected cluster id
-            # (ignore the .batch and .extern lines).
-            if job_id not in job_map:
-                continue
-
-            job = job_map[ClusterJobId(job_id)]
-            assert job.run_script_path is not None
-
-            # extract actual exit code from the ExitCode field
-            exit_code = exit_code.partition(":")[0]
-
-            # Job is considered failed if it has return code 1 or if the state indicates
-            # a failure.
-            if exit_code == 1 or (
-                state in self.job_state_good and not self.job_state_good[state]
-            ):
+        job_statuses = extract_job_status_from_sacct_output(output)
+
+        for job_id, status in job_statuses.items():
+            if job_id in job_map and not status.is_okay():
+                job = job_map[ClusterJobId(job_id)]
+                assert job.run_script_path is not None
+
                 # write hostname to job (it is used in the error message)
-                job.hostname = node_list
+                job.hostname = status.node_list
 
                 # read error message from stderr output file
                 stderr_file = pathlib.Path(job.run_script_path).with_suffix(".err")
                 error_output = stderr_file.read_text()
```

Comment:
I'm wondering if it's good to pull the full error file, or if some additional parsing should be done. At least for my jobs, the error log contains a lot of other stuff (because applications write to stderr). 

Generically handling this parsing will be very difficult though.

Reply:
I didn't think about that.  I agree that parsing for the error will be very difficult.  How about we simply limit it to the last *n* lines of the error log?  Not sure what the best value for *n* would be, maybe 5 would be enough for most cases?  One can always check the actual error log file for the full output if needed.

Reply:
I capped it to 5 lines now.  On a quick test, it looked like this, which I think is enough information in most cases:
```
 Job 19 on hostname galvani-cn001 failed with error:                                                                                                                                                                                                                                                                          
Job failed with state FAILED / exit code 1. Error output (last 5 lines):                                                                                                                                                                                                                                                      
-----                                                                                                                                                                                                                                                                                                                         
  File "/mnt/qb/home/martius/mot456/cluster_utils/examples/basic/main.py", line 48, in <module>
    raise ValueError("5 percent of all jobs die early for testing")
ValueError: 5 percent of all jobs die early for testing
srun: error: galvani-cn001: task 0: Exited with exit code 1
srun: Terminating StepId=295422.0
-----
```